File: examples/conduit/frontend/swiftui/Services/APIService.swift
// Conduit Example App -- SwiftUI API Service
// URLSession-based HTTP client for the Conduit REST API.

import Foundation

enum APIError: LocalizedError {
    case httpError(Int, String)
    case networkError(Error)
    case decodingError(Error)

    var errorDescription: String? {
        switch self {
        case .httpError(_, let message): return message
        case .networkError(let error): return error.localizedDescription
        case .decodingError(let error): return "Decoding error: \(error.localizedDescription)"
        }
    }
}

@MainActor
class APIService: ObservableObject {
    static let shared = APIService()

    private let baseURL = "http://localhost:3000"
    private let session = URLSession.shared
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    @Published var token: String? = nil
    @Published var currentUser: User? = nil

    var isAuthenticated: Bool { token != nil }

    // MARK: - Generic Request

    private func request<T: Decodable>(_ method: String, path: String, body: Any? = nil) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw APIError.httpError(0, "Invalid URL")
        }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = token {
            req.setValue("Token \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body = body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await session.data(for: req)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.httpError(0, "Invalid response")
        }

        if httpResponse.statusCode >= 400 {
            if let errorRes = try? decoder.decode(ErrorResponse.self, from: data) {
                throw APIError.httpError(httpResponse.statusCode, errorRes.errors.body.joined(separator: ", "))
            }
            throw APIError.httpError(httpResponse.statusCode, "HTTP \(httpResponse.statusCode)")
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decodingError(error)
        }
    }

    // MARK: - Auth

    func register(username: String, email: String, password: String) async throws -> User {
        let body: [String: Any] = ["user": ["username": username, "email": email, "password": password]]
        let res: UserResponse = try await request("POST", path: "/api/users", body: body)
        token = res.user.token
        currentUser = res.user
        return res.user
    }

    func login(email: String, password: String) async throws -> User {
        let body: [String: Any] = ["user": ["email": email, "password": password]]
        let res: UserResponse = try await request("POST", path: "/api/users/login", body: body)
        token = res.user.token
        currentUser = res.user
        return res.user
    }

    func logout() {
        token = nil
        currentUser = nil
    }

    // MARK: - Profile

    func getProfile(username: String) async throws -> Profile {
        let res: ProfileResponse = try await request("GET", path: "/api/profiles/\(username)")
        return res.profile
    }

    func updateProfile(bio: String?, image: String?) async throws -> User {
        var userDict: [String: String] = [:]
        if let bio = bio { userDict["bio"] = bio }
        if let image = image { userDict["image"] = image }
        let res: UserResponse = try await request("PUT", path: "/api/user", body: ["user": userDict])
        currentUser = res.user
        return res.user
    }

    // MARK: - Articles

    func getArticles() async throws -> [Article] {
        let res: ArticlesResponse = try await request("GET", path: "/api/articles")
        return res.articles
    }

    func getArticle(slug: String) async throws -> Article {
        let res: ArticleResponse = try await request("GET", path: "/api/articles/\(slug)")
        return res.article
    }

    func createArticle(title: String, description: String, body: String, tagList: [String]?) async throws -> Article {
        var articleDict: [String: Any] = ["title": title, "description": description, "body": body]
        if let tags = tagList { articleDict["tagList"] = tags }
        let res: ArticleResponse = try await request("POST", path: "/api/articles", body: ["article": articleDict])
        return res.article
    }

    func deleteArticle(slug: String) async throws {
        let _: [String: String] = try await request("DELETE", path: "/api/articles/\(slug)")
    }

    // MARK: - Comments

    func getComments(slug: String) async throws -> [Comment] {
        let res: CommentsResponse = try await request("GET", path: "/api/articles/\(slug)/comments")
        return res.comments
    }

    func createComment(slug: String, body: String) async throws -> Comment {
        let res: CommentResponse = try await request("POST", path: "/api/articles/\(slug)/comments", body: ["comment": ["body": body]])
        return res.comment
    }

    func deleteComment(slug: String, commentId: String) async throws {
        let _: [String: String] = try await request("DELETE", path: "/api/articles/\(slug)/comments/\(commentId)")
    }

    // MARK: - Social

    func follow(username: String) async throws -> Profile {
        let res: ProfileResponse = try await request("POST", path: "/api/profiles/\(username)/follow")
        return res.profile
    }

    func unfollow(username: String) async throws -> Profile {
        let res: ProfileResponse = try await request("DELETE", path: "/api/profiles/\(username)/follow")
        return res.profile
    }

    func favorite(slug: String) async throws -> Article {
        let res: ArticleResponse = try await request("POST", path: "/api/articles/\(slug)/favorite")
        return res.article
    }

    func unfavorite(slug: String) async throws -> Article {
        let res: ArticleResponse = try await request("DELETE", path: "/api/articles/\(slug)/favorite")
        return res.article
    }

    // MARK: - Tags

    func getTags() async throws -> [String] {
        let res: TagsResponse = try await request("GET", path: "/api/tags")
        return res.tags
    }
}

File: examples/conduit/frontend/nativescript/services/api-service.ts
// Conduit Example App -- NativeScript API Service
// HTTP client using NativeScript's Http module for the Conduit REST API.

import { Http, HttpResponse } from '@nativescript/core';

const BASE_URL = 'http://localhost:3000';

// Models
export interface User {
  username: string;
  email: string;
  token: string;
  bio?: string;
  image?: string;
}

export interface Profile {
  username: string;
  bio?: string;
  image?: string;
  following: boolean;
}

export interface Article {
  slug: string;
  title: string;
  description: string;
  body: string;
  tagList: string[];
  createdAt: string;
  updatedAt: string;
  favorited: boolean;
  favoritesCount: number;
  author: Profile;
}

export interface Comment {
  id: string;
  body: string;
  createdAt: string;
  author: Profile;
}

export interface UserResponse { user: User }
export interface ProfileResponse { profile: Profile }
export interface ArticleResponse { article: Article }
export interface ArticlesResponse { articles: Article[]; articlesCount: number }
export interface CommentResponse { comment: Comment }
export interface CommentsResponse { comments: Comment[] }
export interface TagsResponse { tags: string[] }
export interface ErrorResponse { errors: { body: string[] } }

// Token management
let authToken: string | null = null;
let currentUser: User | null = null;

export function setToken(token: string | null): void {
  authToken = token;
}

export function getToken(): string | null {
  return authToken;
}

export function getCurrentUser(): User | null {
  return currentUser;
}

export function isAuthenticated(): boolean {
  return authToken !== null;
}

export function logout(): void {
  authToken = null;
  currentUser = null;
}

// Generic request helper
async function request<T>(method: string, path: string, body?: unknown): Promise<T> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (authToken) {
    headers['Authorization'] = `Token ${authToken}`;
  }

  const options: any = {
    url: `${BASE_URL}${path}`,
    method,
    headers,
    timeout: 15000,
  };

  if (body) {
    options.content = JSON.stringify(body);
  }

  let response: HttpResponse;
  try {
    response = await Http.request(options);
  } catch (err) {
    throw new Error(err instanceof Error ? err.message : 'Network error');
  }

  const data = response.content?.toJSON?.() ?? JSON.parse(response.content?.toString?.() ?? '{}');

  if (response.statusCode >= 400) {
    const errorRes = data as ErrorResponse;
    const msg = errorRes?.errors?.body?.join(', ') || `HTTP ${response.statusCode}`;
    throw new Error(msg);
  }

  return data as T;
}

// Auth
export async function login(email: string, password: string): Promise<User> {
  const res = await request<UserResponse>('POST', '/api/users/login', {
    user: { email, password },
  });
  authToken = res.user.token;
  currentUser = res.user;
  return res.user;
}

export async function register(username: string, email: string, password: string): Promise<User> {
  const res = await request<UserResponse>('POST', '/api/users', {
    user: { username, email, password },
  });
  authToken = res.user.token;
  currentUser = res.user;
  return res.user;
}

// Profile
export async function getProfile(username: string): Promise<Profile> {
  const res = await request<ProfileResponse>('GET', `/api/profiles/${username}`);
  return res.profile;
}

export async function updateProfile(bio?: string, image?: string): Promise<User> {
  const res = await request<UserResponse>('PUT', '/api/user', {
    user: { bio, image },
  });
  currentUser = res.user;
  return res.user;
}

// Articles
export async function getArticles(): Promise<Article[]> {
  const res = await request<ArticlesResponse>('GET', '/api/articles');
  return res.articles;
}

export async function getArticle(slug: string): Promise<Article> {
  const res = await request<ArticleResponse>('GET', `/api/articles/${slug}`);
  return res.article;
}

export async function createArticle(
  title: string, description: string, body: string, tagList?: string[]
): Promise<Article> {
  const res = await request<ArticleResponse>('POST', '/api/articles', {
    article: { title, description, body, tagList },
  });
  return res.article;
}

export async function deleteArticle(slug: string): Promise<void> {
  await request<void>('DELETE', `/api/articles/${slug}`);
}

// Comments
export async function getComments(slug: string): Promise<Comment[]> {
  const res = await request<CommentsResponse>('GET', `/api/articles/${slug}/comments`);
  return res.comments;
}

export async function createComment(slug: string, body: string): Promise<Comment> {
  const res = await request<CommentResponse>('POST', `/api/articles/${slug}/comments`, {
    comment: { body },
  });
  return res.comment;
}

export async function deleteComment(slug: string, commentId: string): Promise<void> {
  await request<void>('DELETE', `/api/articles/${slug}/comments/${commentId}`);
}

// Social
export async function follow(username: string): Promise<Profile> {
  const res = await request<ProfileResponse>('POST', `/api/profiles/${username}/follow`);
  return res.profile;
}

export async function unfollow(username: string): Promise<Profile> {
  const res = await request<ProfileResponse>('DELETE', `/api/profiles/${username}/follow`);
  return res.profile;
}

export async function favorite(slug: string): Promise<Article> {
  const res = await request<ArticleResponse>('POST', `/api/articles/${slug}/favorite`);
  return res.article;
}

export async function unfavorite(slug: string): Promise<Article> {
  const res = await request<ArticleResponse>('DELETE', `/api/articles/${slug}/favorite`);
  return res.article;
}

// Tags
export async function getTags(): Promise<string[]> {
  const res = await request<TagsResponse>('GET', '/api/tags');
  return res.tags;
}
